wagtail/admin/rich_text/converters/html_to_contentstate.py

```diff
@@ -315,9 +315,11 @@ def handle_endtag(self, name):
             element_handler.handle_endtag(name, self.state, self.contentstate)
 
     def handle_data(self, content):
-        # normalise whitespace sequences to a single space
+        # normalise whitespace sequences to a single space unless it is contained in <pre> tag, in
+        # which case leave it alone
         # This is in line with https://www.w3.org/TR/html4/struct/text.html#h-9.1
-        content = re.sub(WHITESPACE_RE, ' ', content)
+        if not (content.startswith('<pre>') and content.endswith('</pre>')):
```

Comment:
This seems a little odd - the purpose of [`HTMLParser`](https://docs.python.org/3/library/html.parser.html) is to translate the string representation of HTML into a structure we can iterate over, so I'd expect it to process an element like

    <pre>if x &lt; 5:</pre>

by

* calling `handle_starttag` with the tag name `'pre'`
* calling `handle_data` with the content `"if x < 5:"` (note that the `<` is unescaped at this point, since it's giving us the tag content in plain text form, rather than the HTML serialisation of it)
* calling `handle_endtag` with the tag name `'pre'`

The only time I'd expect to see `<pre>...</pre>` within `handle_data` would be if that's part of the actual document text, i.e. the HTML is something like

    <p>&lt;pre&gt; and &lt;/pre&gt; are used to indicate preformatted text.</p>

@imansyed000, please can you give some details of how you tested this, so I can investigate further?

Reply:
I think I've probably misunderstood the issue, which I thought was the latter description.